### PR TITLE
Update pybind11 to 2.7.1.

### PR DESCRIPTION
**Pull request by clalancette**
This is the version that is shipped in Ubuntu 22.04.
While we are in here, also make it so we only build pybind11
if we need it.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

Still in draft because I'm pretty sure we need some other PRs to go along with this, and I haven't tested on Windows or Windows Debug yet.

### CODE REVIEW

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15629/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10310/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13297/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15846/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/15847/)

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-12-16-2021/23541/1

**Comment by jspricke** [comment]
I guess pybind11 in Ubuntu 22.04 will rather be 2.8.1 or even later versions, see:
https://tracker.debian.org/pkg/pybind11
https://people.canonical.com/~ubuntu-archive/proposed-migration/update_excuses.html#pybind11

**Comment by clalancette** [comment]
> I guess pybind11 in Ubuntu 22.04 will rather be 2.8.1 or even later versions, see:

Thanks for the update.  As soon as I figure out the Windows Debug error, I think we should still bump to 2.7.1 for now.  Once Jammy updates, we can look at updating again.

**Comment by clalancette** [comment]
All right.  With a lot of help from @sloretz , I *think* I have Window Debug fixed here.  Here is another round of CI to see if that is the case:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15893/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10568/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13557/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16199/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/16200/)

**Comment by jspricke** [comment]
FYI: https://tracker.debian.org/news/1292408/accepted-pybind11-290-1-source-into-unstable/

**Comment by clalancette** [comment]
@ros-pull-request-builder retest this please

**Comment by clalancette** [comment]
All right, Windows (really CMake 3.22) had one more warning.  I've now fixed that in the latest.  Otherwise, this looks pretty good.  So I'm going to mark this as ready for review and run one more CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15897/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10572/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13561/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16207/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/16208/)

**Comment by clalancette** [comment]
Woohoo!  All the failing tests here are already known for Windows, so I think this is ready to go once I get approvals.

File: CMakeLists.txt
Comment (sloretz) [inline CMakeLists.txt:57]: I would recommend using the archive URL over the git one, or at least specifying `GIT_SHALLOW` to avoid downloading all the history of pybind11.

File: CMakeLists.txt
Comment (sloretz) [inline CMakeLists.txt:74]: CMake has a warning about patching being non-intuitive when using the `git` download method. I think this would only be an issue if Pybind11 changed what the `v2.7.1` tag pointed at, or if someone changed the git tag locally and rebuilt pybind11_vendor.

PATCH_COMMAND <cmd>...

>    Note that it can be quite difficult to define an appropriate patch command that performs robustly, especially for download methods such as git where changing the GIT_TAG will not discard changes from a previous patch, but the patch command will be called again after updating to the new tag.

File: pybind11-2.7.1-fix-windows-debug.patch
Comment (sloretz) [inline pybind11-2.7.1-fix-windows-debug.patch:41]: Where does this warning happen, and what does it look like?

Comment (clalancette) [reply pybind11-2.7.1-fix-windows-debug.patch:41]: > Where does this warning happen, and what does it look like?

It happens when "building" the `pybind11_vendor` package, at https://github.com/pybind/pybind11/blob/v2.7.1/CMakeLists.txt#L98-L99.  The warning looks like this:

```
Policy CMP0127 is not set: cmake_dependent_option() supports full Condition
Syntax.  Run "cmake --help-policy CMP0127" for policy details.  Use the
cmake_policy command to set the policy and suppress this warning.
```

From my reading of https://cmake.org/cmake/help/latest/policy/CMP0127.html , I don't think we would need *any* changes to this particular line of code, but that doesn't seem to be the case.  No amount of quoting I did on that line made the warning go away, so I just gave up and set CMP0127 to "OLD".  If you have another idea on how to fix it, I'm happy to try that instead.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt:57]: Unfortunately, we have to use the `git` URL here.

The problem is the PATCH_COMMAND below.  On Windows, we cannot use just `patch.exe`, since that requires Administrator permission.  That means we have to use `git apply-patch`.  But we've run into issues in the past where `git apply-patch` doesn't do the correct thing unless it is executed inside a git repository.  So we have to use the `GIT_REPOSITORY` here.

I guess I could use `GIT_SHALLOW` here, but we don't do it in the myriad other places we use this pattern.  I'm happy to do a follow-up with all of them if you think that is worthwhile.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt:74]: Yes, this is completely broken as far as I can tell.  Right now, if packages using this method ever try to rebuild, they fail to apply the patch and fail the build.  Honestly I think `ExternalProject_add` isn't fit for the purpose we are using it for here, but again this is a problem throughout all of our vendor packages.  I'd love a way to fix it (which probably involves not using `ExternalProject_add`, but that is a larger change).

**Comment by clalancette** [comment]
Another CI with the changes in https://github.com/ros2/rclpy/pull/874 and https://github.com/ros2/rosbag2/pull/945

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15906/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10581/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13570/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16227/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/16228/)

File: CMakeLists.txt
Comment (sloretz) [reply CMakeLists.txt:57]: > I guess I could use GIT_SHALLOW here, but we don't do it in the myriad other places we use this pattern. I'm happy to do a follow-up with all of them if you think that is worthwhile.

Probably not worthwhile on this repo. `time git clone https://github.com/pybind/pybind11.git -b v2.7.1 --depth 1` saves very little time on this repo: 3.6 seconds without `--depth 1` and 3.3 seconds with.

**Review by sloretz — APPROVED** [review]

**Comment by clalancette** [comment]
All right, I have approvals on this and the two dependent PRs.  CI is green except for Windows, but those are known failures in rosbag2.  So I'm going to go ahead and merge this whole set.  Thanks for the reviews @sloretz !

**Pull request merged**
